feature-pitches/002-file-based-workflows/002-file-based-workflows.md

```diff
@@ -0,0 +1,112 @@
+
+This would put workflows in front of more users, increasing awareness and hopefully also adoption. 
+
+### UX Note: Re-runnability & Pre-existing Outputs
+
+A common frustration users have with current Renku workflows is that the `workflow` & `run` (mostly `run`) commands errors when you run a workflow command that writes an output file that already exists.
+
+Users do this frequently during development: run script `a`, generate file `b`. Change script `a`, rerun, generate a new b. Great, now you're happy with script `a`, so `renku run script_a` one last time to record it. But now Renku complains that no outputs were generated (because file `b` already existed, so nothing was detected)! Users *can* avoid this by telling Renku explicitly that the output is file `b`, but that's more typing. More commonly, users end up deleting output files before every `renku run`, which they- very reasonably- don't like doing.
+
+The file-based workflow feature should solve this. The workflow should read the specified output in the workflow file, rather than detecting changed files. This removes the need for the user to delete the file so that Renku can detect its creation.
+
+
+## 🐰 Rabbit Holes
+
+### Venturing into the Metadata Realm
+
+This pitch will require a substantial effort to make the metadata all play nice. Without diving into implementation details, here are a few high-level requirments that the metadata must meet in order to provide a good user experience:
+
+* The workflows defined in files should be fully represented in the Renku KG metadata.
+* However, the workflow file itself should be the single source of truth for the workflow definition (not the KG metadata).
+* File based workflows can only be edited via the file. These workflows cannot be edited via the CLI or (in the future) somewhere in the RenkuLab UI.
+* In terms of versioning a file-based workflow, the metadata should contain metadata about the file-based workflow at each execution. There is no need to maintain a history of the file-based workflow at commits where the workflow was modified but not executed.
+* However, the KG must be aware of the latest state of the workflow file (even un-executed or not committed), because other parts of RenkuLab will be pulling information about the workflow via the KG (for example, the workflow browser will query the KG for the file-based workflow definition, and this needs to be up to date with the current state of the file so as not to confuse the user, even if the workflow file has not been executed since it was last modified).
```

Comment:
What you do mean by `not committed` in the first sentence? Is it a typo or does it mean that KG should be updated in real-time?

Reply:
Not a typo... @Panaetius live-updating the KG between commits was part of the implementation proposal we talked through, right? Or did I misunderstand?

I was walking through an example use case of why live-updating to the KG would be important, but then concluded that it would _not_ be so important, so maybe we could remove this requirement? The example I am thinking about is the workflow browser that is being built right now. If a user is writing a workflow in a session, and then looks at the workflow browser to see their workflow, it could be confusing if the workflow browser showed an out of date workflow. But perhaps the user would understand that the workflow browser won't update until they commit?

Reply:
That sounds very complicated imho - maybe if we make it clear that the workflow state corresponds to a specific state of the repository it would be enough of a clue for most users?

Reply:
I don't think we want live updating with uncommitted stuff. but we want to at least process the latest commit (if not executed) so a user doesn't go "Why don't I see my changes in the UI that I committed and push?"

Reply:
So "not committed" should be dropped from that sentence.

Reply:
Ok, fixed.